connectors/sources/github.py

```diff
@@ -0,0 +1,252 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""GitHub source module responsible to fetch documents from GitHub Cloud and Server."""
+import json
+from functools import cached_property
+
+import aiohttp
+from gidgethub import BadRequest
+from gidgethub.aiohttp import GitHubAPI
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource, ConfigurableFieldValueError
+from connectors.utils import CancellableSleeps, RetryStrategy, iso_utc, retryable
+
+RETRIES = 3
+RETRY_INTERVAL = 2
+
+USER_API = "https://api.github.com/user"
+ENDPOINTS = {
+    "ALL_REPOS": "/user/repos?type=all",
+    "REPO": "repos/{repo_name}",
+    "USER": "/user",
+}
+
+
+class GitHubClient:
+    """GitHub Client"""
+
+    def __init__(self, configuration):
+        """Setup the GitHub client.
+
+        Args:
+            configuration (DataSourceConfiguration): Instance of DataSourceConfiguration class.
+        """
+
+        self.session = None
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self.repos = self.configuration["repositories"]
+        self.github_token = self.configuration["github_token"]
+        self.headers = {"Authorization": f"Bearer {self.github_token}"}
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def execute_getitem(self, request):
+        """Execute request using getitem method of GitHubAPI.
+
+        Args:
+            request (str): API for the request
+
+        Returns:
+            list: Response of the request
+        """
+        return await self._get_session.getitem(url=request)
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def execute_request(self, method, url):
+        """Execute request using _request method of GitHubAPI.
+
+        Args:
+            request (str): API for the request
+
+        Returns:
+            list: Response of the request
+        """
+        status_code, response_headers, data = await self._get_session._request(
+            method="get", url=USER_API, headers=self.headers
+        )
+        if status_code == 200:
+            return response_headers
+        data = data.decode("utf-8")
+        message = json.loads(data)
+        raise Exception(
+            f"Something went wrong while executing request Exception: {message['message']}"
+        )
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def execute_getiter(self, request):
```

Comment:
```suggestion
    async def get_resource(self, resource):
```

suggestion: I think we can call this method `get_resource`, `yield_resource` or `resource_generator` to make it more explicit what's returned by that method.